File: src/main/java/com/crealytics/spark/excel/PlainNumberFormat.java
package com.crealytics.spark.excel;

import java.math.BigDecimal;
import java.text.FieldPosition;
import java.text.Format;
import java.text.ParsePosition;


/**
 * A format that formats a double as a plain string without rounding and scientific notation.
 * All other operations are unsupported.
 * @see <a href="https://github.com/apache/poi/blob/REL_4_1_2/src/java/org/apache/poi/ss/usermodel/ExcelGeneralNumberFormat.java">
 * ExcelGeneralNumberFormat</a> and <a href="https://github.com/apache/poi/blob/REL_4_1_2/src/java/org/apache/poi/ss/usermodel/DataFormatter.java#L1171">
 * SSNFormat</a> from Apache POI.
 **/
public class PlainNumberFormat extends Format {

    private static final long serialVersionUID = 1L;

    private static final Format INSTANCE = new PlainNumberFormat();

    private PlainNumberFormat() {
        // enforce singleton
    }

    public static Format getInstance() {
        return INSTANCE;
    }

    @Override
    public StringBuffer format(Object number, StringBuffer toAppendTo, FieldPosition pos) {
        return toAppendTo.append(new BigDecimal(number.toString()).toPlainString());
    }

    @Override
    public Object parseObject(String source, ParsePosition pos) {
        throw new UnsupportedOperationException();
    }

}

File: src/main/scala/com/crealytics/spark/excel/PlainNumberFormat.scala
/*
 * Copyright 2023 Martin Mauch (@nightscape)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crealytics.spark.excel

import java.math.BigDecimal
import java.text.FieldPosition
import java.text.Format
import java.text.ParsePosition

/** A format that formats a double as a plain string without rounding and scientific notation. All other operations are
  * unsupported.
  * @see
  *   [[org.apache.poi.ss.usermodel.ExcelGeneralNumberFormat]] and SSNFormat from
  *   [[org.apache.poi.ss.usermodel.DataFormatter]] from Apache POI.
  */
object PlainNumberFormat extends Format {

  override def format(number: AnyRef, toAppendTo: StringBuffer, pos: FieldPosition): StringBuffer =
    toAppendTo.append(new BigDecimal(number.toString).toPlainString)

  override def parseObject(source: String, pos: ParsePosition): AnyRef =
    throw new UnsupportedOperationException()
}
